internals/feature_helpers.py

```diff
@@ -480,37 +482,48 @@ def get_by_ids(feature_ids: list[int],
                      for f in cached_features.values()
                      if f is not None and f.get('id')}
 
-  for feature_id in feature_ids:
-    if result_dict.get(feature_id) is None:
-      futures_by_id[feature_id] = FeatureEntry.get_by_id_async(feature_id)
+  needed_ids = [fid for fid in feature_ids if fid not in result_dict]
+
+  if needed_ids:
+    # Create ndb.Key objects for the batch operation
+    needed_keys = [ndb.Key(FeatureEntry, fid) for fid in needed_ids]
+
+    # Get all features and stages asynchronously.
+    unformatted_features_future: list[ndb.Future] = ndb.get_multi_async(needed_keys)
+    stages_futures: list[ndb.Future] = []
```

Comment:
Question:

Looking at the current database, what's the worst case scenario of how many stages will need to be fetched?

If it's a lot, we might want to do batches of what you were kinda doing in the previous PR:

```
BATCH_SIZE = 30
...

stages_futures = []
for i in range(0, len(ids), BATCH_SIZE):
  batch_ids = ids[i:i + BATCH_SIZE]
  # fetch_async returns a Future object immediately
  stages_futures.append(
     Stage.query(Stage.feature_id.IN(batch_ids), Stage.archived == False)
      .fetch_async(None))
```

If the number is low, keeping it as-is is good. But if we are talking 3k feature IDs, batching takes the number of requests down from 3k to 100.

Reply:
I do think that the suggestion might make a slightly more optimal set of queries, but I don't think the number of entities that we query for is high enough to worry too much about it. If we find the site is starting to feel bogged-down, this might be the right move.

Do you think I should create an issue regarding this possible optimization?

Reply:
We can still merge this in the meantime.